File: op-private-interop/projection/commit.go
package projection

import (
	"encoding/binary"
	"fmt"
	"slices"

	"github.com/ethereum-optimism/optimism/op-private-interop/wire"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"
)

// Domain tags of the sp1-private-projection-v1 statement (§C.3.1). Byte-identical in Kona.
const (
	ConfigDomain        = "optimism.private-projection-config.v1\x00"
	PrivateConfigDomain = "optimism.private-config.v1\x00"
	DepSetDomain        = "optimism.private-dependency-set.v1\x00"
	OutputsDomain       = "optimism.private-outputs.v1\x00"
	MessagesDomain      = "optimism.private-messages.v1\x00"
)

// Message leaf kinds (§C.3.7). Kind 0x03 (generic EventReplayer) is not defined in v1.
const (
	MessageKindInit byte = 0x01 // export: SentMessage
	MessageKindExec byte = 0x02 // import: ExecutingMessage
)

// commitmentTop is the unwrapped tree top: zero for no leaves, else the RecordsRoot-shaped tree
// (nodes H(0x01 ‖ left ‖ right), duplicating the last node at odd levels).
func commitmentTop(leaves []common.Hash) common.Hash {
	if len(leaves) == 0 {
		return common.Hash{}
	}
	level := slices.Clone(leaves)
	for len(level) > 1 {
		next := make([]common.Hash, 0, (len(level)+1)/2)
		for i := 0; i < len(level); i += 2 {
			right := level[min(i+1, len(level)-1)]
			next = append(next, crypto.Keccak256Hash([]byte{1}, level[i][:], right[:]))
		}
		level = next
	}
	return level[0]
}

func wrapCommitment(domain string, n uint64, top common.Hash) common.Hash {
	var count [8]byte
	binary.BigEndian.PutUint64(count[:], n)
	return crypto.Keccak256Hash([]byte(domain), count[:], top[:])
}

// CommitmentRoot is keccak256(domain ‖ u64be(n) ‖ top) over the §C.3.5 tree. The leaf count is
// committed separately so odd-level duplication cannot alias different lengths.
func CommitmentRoot(domain string, leaves []common.Hash) common.Hash {
	return wrapCommitment(domain, uint64(len(leaves)), commitmentTop(leaves))
}

// CommitmentProof returns the bottom-up siblings of leaves[index]. At a level where the node is
// the last element of an odd-length level its sibling is itself, and no element is emitted.
func CommitmentProof(leaves []common.Hash, index int) ([]common.Hash, error) {
	if index < 0 || index >= len(leaves) {
		return nil, fmt.Errorf("commitment proof index %d out of range [0,%d)", index, len(leaves))
	}
	var siblings []common.Hash
	level := slices.Clone(leaves)
	for len(level) > 1 {
		if sib := index ^ 1; sib < len(level) {
			siblings = append(siblings, level[sib])
		}
		next := make([]common.Hash, 0, (len(level)+1)/2)
		for i := 0; i < len(level); i += 2 {
			right := level[min(i+1, len(level)-1)]
			next = append(next, crypto.Keccak256Hash([]byte{1}, level[i][:], right[:]))
		}
		level = next
		index /= 2
	}
	return siblings, nil
}

// VerifyCommitmentProof recomputes the root from one leaf. Level lengths are derived from n, the
// odd-level rule is applied identically, and every sibling must be consumed.
func VerifyCommitmentProof(domain string, n, index uint64, leaf common.Hash, siblings []common.Hash, root common.Hash) bool {
	if n == 0 || index >= n {
		return false
	}
	node, width, used := leaf, n, 0
	for width > 1 {
		if index%2 == 1 {
			if used >= len(siblings) {
				return false
			}
			node = crypto.Keccak256Hash([]byte{1}, siblings[used][:], node[:])
			used++
		} else if index+1 < width {
			if used >= len(siblings) {
				return false
			}
			node = crypto.Keccak256Hash([]byte{1}, node[:], siblings[used][:])
			used++
		} else {
			node = crypto.Keccak256Hash([]byte{1}, node[:], node[:])
		}
		index /= 2
		width = (width + 1) / 2
	}
	return used == len(siblings) && wrapCommitment(domain, n, node) == root
}

// OutputLeaf = keccak256(0x00 ‖ u64be(blockNumber) ‖ outputRoot).
func OutputLeaf(blockNumber uint64, outputRoot common.Hash) common.Hash {
	var n [8]byte
	binary.BigEndian.PutUint64(n[:], blockNumber)
	return crypto.Keccak256Hash([]byte{0}, n[:], outputRoot[:])
}

// MessageLeaf = keccak256(0x00 ‖ u64be(blockNumber) ‖ u32be(renderedIndex) ‖ u8(kind) ‖ messageHash).
func MessageLeaf(blockNumber uint64, renderedIndex uint32, kind byte, messageHash common.Hash) common.Hash {
	var n [8]byte
	var r [4]byte
	binary.BigEndian.PutUint64(n[:], blockNumber)
	binary.BigEndian.PutUint32(r[:], renderedIndex)
	return crypto.Keccak256Hash([]byte{0}, n[:], r[:], []byte{kind}, messageHash[:])
}

// ExportMessageHash is the interop payload hash keccak256(topic0 ‖ … ‖ topic3 ‖ data) of the
// SentMessage log the replay messenger emits for m.
func ExportMessageHash(m *wire.SentMessage) common.Hash {
	topics, data := wire.SentMessageLog(m)
	h := crypto.NewKeccakState()
	for _, t := range topics {
		_, _ = h.Write(t[:])
	}
	_, _ = h.Write(data)
	var out common.Hash
	_, _ = h.Read(out[:])
	return out
}

// ImportMessageHash = keccak256(identifierWords(160) ‖ payloadHash(32)): the 192 argument bytes of
// validateMessage(Identifier,bytes32), equal by ABI to ExecutingMessage's data ‖ topics[1].
func ImportMessageHash(args [192]byte) common.Hash {
	return crypto.Keccak256Hash(args[:])
}

File: rust/kona/crates/protocol/protocol/src/projection/commit.rs
//! Domain-parameterised commitment tree and the output and message leaves of the
//! sound private-projection statement. Twin of Go `op-private-interop/projection/commit.go`.

use alloc::vec::Vec;
use alloy_primitives::{B256, keccak256};

use super::replaySentMessageCall;

/// Domain of the per-block private output commitment (`outputsRoot`).
pub const OUTPUTS_DOMAIN: &[u8] = b"optimism.private-outputs.v1\0";
/// Domain of the rendered-message commitment (`messagesRoot`).
pub const MESSAGES_DOMAIN: &[u8] = b"optimism.private-messages.v1\0";

/// Message leaf kind of a rendered message.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum MessageKind {
    /// Export: a `SentMessage` from `L2ToL2CrossDomainMessenger`.
    Init = 0x01,
    /// Import: an `ExecutingMessage` from `CrossL2Inbox`.
    Exec = 0x02,
}

fn node(left: &B256, right: &B256) -> B256 {
    let mut buf = [0u8; 65];
    buf[0] = 1;
    buf[1..33].copy_from_slice(left.as_slice());
    buf[33..].copy_from_slice(right.as_slice());
    keccak256(buf)
}

fn finish(domain: &[u8], n: u64, top: &B256) -> B256 {
    let mut buf = Vec::with_capacity(domain.len() + 40);
    buf.extend_from_slice(domain);
    buf.extend_from_slice(&n.to_be_bytes());
    buf.extend_from_slice(top.as_slice());
    keccak256(buf)
}

/// Count-bound Merkle root over ordered leaves. The last node of an odd level is
/// paired with itself; the empty tree has an all-zero top.
pub fn commitment_root(domain: &[u8], leaves: &[B256]) -> B256 {
    let top = if leaves.is_empty() {
        B256::ZERO
    } else {
        let mut level = leaves.to_vec();
        while level.len() > 1 {
            level = level.chunks(2).map(|p| node(&p[0], p.get(1).unwrap_or(&p[0]))).collect();
        }
        level[0]
    };
    finish(domain, leaves.len() as u64, &top)
}

/// Bottom-up siblings for `leaves[index]`. A node that is the last element of an
/// odd-length level is paired with itself and emits no sibling. `None` if out of range.
pub fn commitment_proof(leaves: &[B256], index: usize) -> Option<Vec<B256>> {
    if index >= leaves.len() {
        return None;
    }
    let mut siblings = Vec::new();
    let mut level = leaves.to_vec();
    let mut idx = index;
    while level.len() > 1 {
        let odd_last = level.len() % 2 == 1 && idx == level.len() - 1;
        if !odd_last {
            siblings.push(level[idx ^ 1]);
        }
        level = level.chunks(2).map(|p| node(&p[0], p.get(1).unwrap_or(&p[0]))).collect();
        idx /= 2;
    }
    Some(siblings)
}

/// Verify an inclusion proof produced by [`commitment_proof`] against a root from
/// [`commitment_root`]. Level lengths are recomputed from `n`; every sibling must be used.
pub fn verify_commitment_proof(
    domain: &[u8],
    n: u64,
    index: u64,
    leaf: B256,
    siblings: &[B256],
    root: B256,
) -> bool {
    if index >= n {
        return false;
    }
    let mut len = n;
    let mut idx = index;
    let mut acc = leaf;
    let mut rest = siblings.iter();
    while len > 1 {
        if len % 2 == 1 && idx == len - 1 {
            acc = node(&acc, &acc);
        } else {
            let Some(sibling) = rest.next() else {
                return false;
            };
            acc = if idx.is_multiple_of(2) { node(&acc, sibling) } else { node(sibling, &acc) };
        }
        idx /= 2;
        len = len.div_ceil(2);
    }
    rest.next().is_none() && finish(domain, n, &acc) == root
}

/// `keccak256(0x00 ‖ u64be(blockNumber) ‖ outputRoot)`.
pub fn output_leaf(block_number: u64, output_root: B256) -> B256 {
    let mut buf = [0u8; 41];
    buf[1..9].copy_from_slice(&block_number.to_be_bytes());
    buf[9..].copy_from_slice(output_root.as_slice());
    keccak256(buf)
}

/// `keccak256(0x00 ‖ u64be(blockNumber) ‖ u32be(renderedIndex) ‖ u8(kind) ‖ messageHash)`.
pub fn message_leaf(
    block_number: u64,
    rendered_index: u32,
    kind: MessageKind,
    message_hash: B256,
) -> B256 {
    let mut buf = [0u8; 46];
    buf[1..9].copy_from_slice(&block_number.to_be_bytes());
    buf[9..13].copy_from_slice(&rendered_index.to_be_bytes());
    buf[13] = kind as u8;
    buf[14..].copy_from_slice(message_hash.as_slice());
    keccak256(buf)
}

/// Interop payload hash of the `SentMessage` log that `replaySentMessage` emits:
/// `keccak256(topic0 ‖ topic1 ‖ topic2 ‖ topic3 ‖ abi.encode(sender, message))`.
pub fn export_message_hash(m: &replaySentMessageCall) -> B256 {
    let (topics, data) = super::render::sent_message_log(m);
    let mut buf = Vec::with_capacity(128 + data.len());
    for t in &topics {
        buf.extend_from_slice(t.as_slice());
    }
    buf.extend_from_slice(&data);
    keccak256(buf)
}

/// `keccak256(identifierWords ‖ payloadHash)`: the 192 argument bytes of
/// `validateMessage(Identifier,bytes32)` calldata (calldata without its selector).
pub fn import_message_hash(args: &[u8; 192]) -> B256 {
    keccak256(args)
}
